### PR TITLE
Add component tests for collection

### PR DESCRIPTION
**Pull request by WStechura**
Add integration tests, used in previous collection. Most of the changes is just a copy-paste of an old test framework, however some adoption was required.
The rests are being run by `run.py` script within `tests/integration` folder. Upon detection `localhost` as one of the platforms parametrers, deployment on local machine is performed. Unfortunately, it does not work for Windows platforms as Ansible does work with this platform as main node.
The tests uses artifical installer, residing in `tests/integration/installers` directory, which is combined with `uninstall.sh` and `oneagentctl` scripts to perform seamless deployment.
Test directory resides in `tests/integration/tests_dir` which is removed after test is performed

### CODE REVIEW

File: .github/actions/setup-build-environment/action.yaml [hunk 1]
```diff
@@ -23,4 +23,4 @@ runs:
         python-version: '3.10'
     - name: Install dependencies
       shell: bash
-      run: pip install ansible
+      run: pip install venv ansible
```
Comment (WStechura) [inline .github/actions/setup-build-environment/action.yaml:26]: Change to virtualenv

File: .github/actions/run-tests/action.yaml [hunk 2]
```diff
@@ -24,3 +24,15 @@ runs:
...
       shell: bash
       run: pushd ~/.ansible/collections/ansible_collections/dynatrace/oneagent && ansible-test sanity && popd
+    - name: Run integration tests
+      id: integration-tests
+      shell: bash
+      run: python -m venv venv && source venv/bin/activate && cd roles/oneagent/tests/integration && python run.py
+    - name: Upload logs
+      if: failure() && steps.integration-tests.outcome == 'failure'
+      uses: actions/upload-artifact@v4
+      run: echo "Uploading logs"
```
Comment (WStechura) [inline .github/actions/run-tests/action.yaml:34]: Remove

File: .idea/inspectionProfiles/Project_Default.xml
Comment (WStechura) [inline .idea/inspectionProfiles/Project_Default.xml:1]: .idea files should be removed

File: roles/oneagent/README.md [hunk 3]
```diff
@@ -39,21 +39,22 @@ For full list of suitable parameters, see [OneAgent configuration via command-li
...
-| Name | Default | Description
-|-|-|-
-| `oneagent_environment_url` | `-` | The URL of the target Dynatrace environment (see [Direct download from your environment](#direct-download-from-your-environment)).
-| `oneagent_paas_token` | `-` | The [PaaS Token] retrieved from the "Deploy Dynatrace" installer page.
-| `oneagent_local_installer` | `-` | The Path to OneAgent installer stored on the main node.
-| `oneagent_installer_arch` | `-` | Specifies the OneAgent installer architecture.
-| `oneagent_version` | `latest` | The required version of the OneAgent in the `1.199.247.20200714-111723` format. See [Deployment API - GET available versions of OneAgent] for more details.
-| `oneagent_download_dir` | Linux: `$TEMP` or `/tmp`</br>Windows: `%TEMP%` or `C:\Windows\Temp` | Installer download directory. For Linux and AIX, the directory must not contain spaces. Will be created if it does not exist.
-| `oneagent_install_args` | `-`  Dynatrace OneAgent installation parameters defined as a list of items.
+| Name                             | Default | Description
```
Comment (WStechura) [inline roles/oneagent/README.md:42]: Was this change intended?

File: roles/oneagent/defaults/main.yml [hunk 4]
```diff
@@ -31,5 +31,8 @@ oneagent_reboot_host: false
 # Set the need of validating certs during installer download
 oneagent_validate_certs: true
 
+# Set the need of verifying the installer signature
+oneagent_verify_signature: true
```
Comment (WStechura) [inline roles/oneagent/defaults/main.yml:35]: Release notes should be updated regarding this parameter

File: roles/oneagent/tests/integration/run.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+def get_test_args(args: Dict[str, Any]) -> List[str]:
+    test_args = [f"--{USER_KEY}='{args[USER_KEY]}'", f"--{PASS_KEY}='{args[PASS_KEY]}'"]
+    for platform in [p.value for p in DeploymentPlatform]:
+        hosts = ','.join(args[platform])
+        hosts and test_args.append(f"--{platform}={hosts}")
+    return test_args
+
+
+def run_test(test: str, test_args: List[str]) -> bool:
+    test_name = Path(test).stem
```
Comment (WStechura) [inline roles/oneagent/tests/integration/run.py:56]: Add log about what test is being run and it's result

File: roles/oneagent/tests/integration/scripts/command/unix/unix_command_executor.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        return []
+    else:
+        return ["sshpass", "-p", f"{password}", "ssh", "-o", "StrictHostKeyChecking=no", f"{user}@{address}"]
+
+
+def _get_command_and_shell(address: str, user: str, password: str, command: str, *args: str) \
+        -> [Union[List[str], str], bool]:
+    prefix = _get_command_prefix(address, user, password)
+    if not prefix:
+        return " ".join([command, *args]), True
```
Comment (WStechura) [inline roles/oneagent/tests/integration/scripts/command/unix/unix_command_executor.py:18]: Add comment why in case of empty prefix we want to use shell=True

File: roles/oneagent/tests/integration/scripts/server/server.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+def get_latest_agent(system):
+    return get_installer(system, request.args["arch"], "latest")
+
+
+@bp.route("/<system>/default/version/<version>")
+def get_agent_in_version(system, version):
+    return get_installer(system, request.args["arch"], version)
+
+
+# def parse_args():
```
Comment (WStechura) [inline roles/oneagent/tests/integration/scripts/server/server.py:44]: Remove comment

File: roles/oneagent/tests/integration/scripts/technology/ansible/config.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        write_yaml_file(PLAYBOOK_FILE, data)
+
+    def clear_parameters_section(self) -> None:
+        data = read_yaml_file(PLAYBOOK_FILE)
+        data[0][AnsibleConfig.PARAM_SECTION_KEY] = {}
+        write_yaml_file(PLAYBOOK_FILE, data)
+        data = read_yaml_file(INVENTORY_FILE)
+        for platform in DeploymentPlatform:
+            group_data = data["all"]["children"][platform.family()]["children"][platform.value]
+
```
Comment (WStechura) [inline roles/oneagent/tests/integration/scripts/technology/ansible/config.py:107]: Remove space

File: roles/oneagent/tests/integration/scripts/tests/test_resilience.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+DOWNLOAD_FAILED_KEY = "failed_download"
+SIGNATURE_VERIFICATION_FAILED_KEY = "signature_verification_failed"
+VERSION_LOWER_THAN_INSTALLED_KEY = "version_lower_than_installed"
+
+
+def _prepare_test_data(constants, data: Dict[str, str]) -> Dict[str, str]:
+    parsed_data = {}
+    prefix_length = len(constants.VARIABLE_PREFIX)
+    for key, value in data.items():
+        # Remove oneagent_ prefix for Ansible and $ for Puppet
```
Comment (WStechura) [inline roles/oneagent/tests/integration/scripts/tests/test_resilience.py:28]: Puppet logic is no longer needed

**Comment by WStechura** [comment]
Make sure the pylint, ansible-lint and shellcheck are passing

File: .github/actions/run-tests/action.yaml [hunk 10]
```diff
@@ -24,3 +24,15 @@ runs:
     - name: Run sanity test
       shell: bash
       run: pushd ~/.ansible/collections/ansible_collections/dynatrace/oneagent && ansible-test sanity && popd
+    - name: Run integration tests
```
Comment (jakubrak) [inline .github/actions/run-tests/action.yaml:27]: Please rename to component tests or e2e tests. We do not test integration, the tests are isolated completely

**Comment by jakubrak** [comment]
Tests are failing with
```
Run pip install venv ansible
  pip install venv ansible
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    FILES_DIR: roles/oneagent/files
    pythonLocation: /opt/hostedtoolcache/Python/3.10.14/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.10.14/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.10.14/x64
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.10.14/x64
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.10.14/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.10.14/x64/lib
ERROR: Could not find a version that satisfies the requirement venv (from versions: none)
ERROR: No matching distribution found for venv
Error: Process completed with exit code 1.
```

File: .github/actions/run-tests/action.yaml (hunk 10 above)
Comment (filip-szweda) [inline .github/actions/run-tests/action.yaml:27]: It would be worth to document how to run the tests in readme.

File: roles/oneagent/tests/conmponent/scripts/util/constant_key.py [hunk 11]
```diff
@@ -0,0 +1,5 @@
+from abc import abstractmethod
+
+
+def constant_key(func):
```
Comment (hparzych) [inline roles/oneagent/tests/conmponent/scripts/util/constant_key.py:4]:
```suggestion
```

File: roles/oneagent/tests/conmponent/scripts/util/test_helpers.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
+import functools
```
Comment (hparzych) [inline roles/oneagent/tests/conmponent/scripts/util/test_helpers.py:1]: refactor the file location to not use `util` directory and name the file with a descriptive name

File: roles/oneagent/tests/conmponent/scripts/util/test_helpers.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+CallableOperation = Callable[[DeploymentPlatform, str, Any], None]
+
+
+def _get_param_by_name(name: str, **kwargs) -> Any:
+    assert name in kwargs, f"No '{name}' parameter in parameters list"
+    return kwargs[name]
+
+
+def disable_for_localhost():
```
Comment (hparzych) [inline roles/oneagent/tests/conmponent/scripts/util/test_helpers.py:22]:
```suggestion
def skip_hosts_for_tests(host_values: list[str]):
```

Comment (hparzych) [reply roles/oneagent/tests/conmponent/scripts/util/test_helpers.py:22]: consider a different approach as this function wouldn't work for addresses such as `127.0.0.1`. why is skipping tests needed for localhost?

File: roles/oneagent/tests/conmponent/scripts/util/test_helpers.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+                if any(host == "localhost" for host in hosts):
+                    return
+            func(*args, **kwargs)
+
+        return platforms_wrapper
+
+    return func_wrapper
+
+
+def enable_for_family(family: str):
```
Comment (hparzych) [inline roles/oneagent/tests/conmponent/scripts/util/test_helpers.py:37]: consider a more descriptive name. what is this function enabling?

File: roles/oneagent/tests/conmponent/scripts/util/constants/common_constants.py [hunk 15]
```diff
@@ -0,0 +1,15 @@
+from pathlib import Path
+
+# is cwd() correct?
```
Comment (hparzych) [inline roles/oneagent/tests/conmponent/scripts/util/constants/common_constants.py:3]: use the path of the script (`__file__`)

File: roles/oneagent/tests/conmponent/scripts/util/test_data_types.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+@dataclass
+class CommandResult:
+    returncode: int
+    stdout: str
+    stderr: str
+
+
+class DeploymentPlatform(Enum):
+    AIX_PPC = "aix_ppc"
```
Comment (hparzych) [inline roles/oneagent/tests/conmponent/scripts/util/test_data_types.py:15]: refactor the code to remove unnecessary abstraction

File: roles/oneagent/tests/conmponent/scripts/tests/conftest.py [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+from technology.ansible.config import AnsibleConfig
+from technology.ansible.runner import AnsibleRunner
+from util.common_utils import prepare_test_dirs
+from util.test_data_types import DeploymentPlatform, PlatformCollection, DeploymentResult
+
+
+USER_KEY = "user"
+PASS_KEY = "password"
+
+UTIL_KEY = "util"
```
Comment (hparzych) [inline roles/oneagent/tests/conmponent/scripts/tests/conftest.py:19]: remove redundant constants

**Review by hparzych — CHANGES_REQUESTED** [review]

**Comment by hparzych** [comment]
Instead of passing `localhost` address for running ansible playbooks, use the appropriate Ansible feature to run playbooks locally

File: roles/oneagent/tests/conmponent/scripts/tests/test_installAndConfig.py [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+CONFIG_INTENDED_TAG = "config_intended_tag"
+CONFIG_INTENDED_PROPERTY = "install_prop2=config_intended_prop"
+
+CTL_TAGS_GETTER = "--get-host-tags"
+CTL_PROPERTIES_GETTER = "--get-host-properties"
+
+INSTALLER_ARGS = [f'--set-host-tag="{INSTALLER_TAG}"', f'--set-host-property="{INSTALLER_PROPERTY}"']
+INSTALLER_PLATFORM_ARGS = [f'--set-host-tag="{PLATFORM_TAG}"', f'--set-host-property="{PLATFORM_PROPERTY}"']
+CONFIG_INSTALL_ARGS = [f'--set-host-tag="{CONFIG_INSTALL_TAG}"', f'--set-host-property="{CONFIG_INSTALL_PROPERTY}"']
+CONFIG_INTENDED_ARGS = [f'--set-host-tag="{CONFIG_INTENDED_TAG}"', f'--set-host-property="{CONFIG_INTENDED_PROPERTY}"']
```
Comment (hparzych) [inline roles/oneagent/tests/conmponent/scripts/tests/test_installAndConfig.py:39]: name `CONFIG_INSTALL_ARGS` and `CONFIG_INTENDED_ARGS` with their purpose as currently, as we talked, it's unclear

File: roles/oneagent/tests/conmponent/run.py [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+
+
+def run_server():
+    logging.info("Running server...")
+    server_path = Path("scripts") / "server" / "server.py"
+    return subprocess.Popen(["python", server_path], env=get_env_vars(),
+                            stdout=subprocess.PIPE, stderr=subprocess.STDOUT, encoding="utf-8")
+
+
+def get_file_content(path: Path):
```
Comment (filip-szweda) [inline roles/oneagent/tests/conmponent/run.py:88]:
```suggestion
def get_file_content(path: Path) -> List[str]:
```

File: roles/oneagent/tests/conmponent/run.py [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+    with open(path, "r") as f:
+        return f.readlines()
+
+
+def replace_tag(source: List[str], old: str, new: str) -> List[str]:
+    return [line.replace(old, new) for line in source]
+
+
+def prepare_installers():
```
Comment (filip-szweda) [inline roles/oneagent/tests/conmponent/run.py:97]:
```suggestion
def prepare_installers()- > None:
```

File: roles/oneagent/tests/conmponent/run.py [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+    timestamp = '{:%Y%m%d-%H%M%S}'.format(datetime.now())
+    # Minimal supported version is 1.199
+    for version in ["1.199.0", "1.300.0"]:
+        full_version = f"{version}.{timestamp}"
+        installer_code = replace_tag(installer_template, version_tag, full_version)
+        with open(INSTALLERS_DIR / f"{installer_partial_name}-{full_version}.sh", "w") as f:
+            f.writelines(installer_code)
+
+
+def prepare_environment():
```
Comment (filip-szweda) [inline roles/oneagent/tests/conmponent/run.py:129]:
```suggestion
def prepare_environment() -> None:
```

File: roles/oneagent/tests/conmponent/run.py [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+
+    for platform in DeploymentPlatform:
+        parser.add_argument(
+            f"--{platform.value}", type=str, nargs="+", default=[], help="List of IPs for specified platform"
+        )
+
+    return vars(parser.parse_args())
+
+
+def main():
```
Comment (filip-szweda) [inline roles/oneagent/tests/conmponent/run.py:157]:
```suggestion
def main() -> bool:
```

File: roles/oneagent/tests/conmponent/run.py [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+    tests_failed = False
+
+    for test in glob.glob(f"{test_path}/test_*.py"):
+        if not run_test(test, test_args):
+            tests_failed = True
+
+    return tests_failed
+
+
+def run_server():
```
Comment (filip-szweda) [inline roles/oneagent/tests/conmponent/run.py:81]:
```suggestion
def run_server() -> Optional[subprocess.Popen]::
```

File: roles/oneagent/tests/conmponent/run.py [hunk 24]
```diff
@@ -0,0 +1,166 @@
...
+
+
+def get_env_vars():
+    # This is required to pass current venv vars down to the subprocess
+    env_vars = os.environ.copy()
+    env_vars.update(TEST_VARS)
+    return env_vars
+
+
+def save_log(out, log_path: Path):
```
Comment (filip-szweda) [inline roles/oneagent/tests/conmponent/run.py:42]:
```suggestion
def save_log(out, log_path: Path) -> None:
```

File: roles/oneagent/tests/conmponent/run.py [hunk 25]
```diff
@@ -0,0 +1,166 @@
...
+
+    def __enter__(self):
+        return self.proc
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.proc.terminate()
+        save_log(self.proc.stdout, LOG_DIR / "server.log")
+
+
+def get_env_vars():
```
Comment (filip-szweda) [inline roles/oneagent/tests/conmponent/run.py:35]:
```suggestion
def get_env_vars() -> Dict[str, str]:
```

File: roles/oneagent/tests/conmponent/run.py (hunk 22 above)
Comment (filip-szweda) [inline roles/oneagent/tests/conmponent/run.py:157]: It would be better to run the tests directly using `pytest`, but I understand that it would take a lot of work to refactor the code.

File: roles/oneagent/tests/component/scripts/command/unix/unix_command_executor.py [hunk 26]
```diff
@@ -0,0 +1,36 @@
+import subprocess
+
+from util.test_data_types import CommandResult
+
+
+def _get_command_prefix(address: str, user: str, password: str) -> list[str]:
+    if address == "localhost":
+        return []
+    else:
```
Comment (filip-szweda) [inline roles/oneagent/tests/component/scripts/command/unix/unix_command_executor.py:9]: No need for else.

File: roles/oneagent/tests/component/scripts/command/command_wrapper.py [hunk 27]
```diff
@@ -0,0 +1,26 @@
+from abc import abstractmethod
+from pathlib import Path
+
+from util.test_data_types import CommandResult
+
+
+class CommandWrapper:
+    @abstractmethod
+    def get_file_content(self, address: str, file: Path) -> CommandResult:
+        pass
```
Comment (filip-szweda) [inline roles/oneagent/tests/component/scripts/command/command_wrapper.py:10]:
```suggestion
        raise NotImplementedError
```

File: roles/oneagent/tests/component/scripts/command/command_wrapper.py [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+
+    @abstractmethod
+    def file_exists(self, address: str, file: Path) -> CommandResult:
+        pass
```
Comment (filip-szweda) [inline roles/oneagent/tests/component/scripts/command/command_wrapper.py:14]:
```suggestion
        raise NotImplementedError
```

File: roles/oneagent/tests/component/scripts/command/command_wrapper.py [hunk 29]
```diff
@@ -0,0 +1,26 @@
...
+
+    @abstractmethod
+    def directory_exists(self, address: str, directory: Path) -> CommandResult:
+        pass
```
Comment (filip-szweda) [inline roles/oneagent/tests/component/scripts/command/command_wrapper.py:18]:
```suggestion
        raise NotImplementedError
```

File: roles/oneagent/tests/component/scripts/command/command_wrapper.py [hunk 30]
```diff
@@ -0,0 +1,26 @@
...
+
+    @abstractmethod
+    def create_directory(self, address: str, directory: Path) -> CommandResult:
+        pass
```
Comment (filip-szweda) [inline roles/oneagent/tests/component/scripts/command/command_wrapper.py:22]:
```suggestion
        raise NotImplementedError
```

File: roles/oneagent/tests/component/scripts/command/command_wrapper.py [hunk 31]
```diff
@@ -0,0 +1,26 @@
...
+
+    @abstractmethod
+    def run_command(self, address: str, command: str, *args: str) -> CommandResult:
+        pass
```
Comment (filip-szweda) [inline roles/oneagent/tests/component/scripts/command/command_wrapper.py:26]:
```suggestion
        raise NotImplementedError
```

File: roles/oneagent/tests/component/scripts/technology/deployment_runner.py [hunk 32]
```diff
@@ -0,0 +1,9 @@
+from abc import abstractmethod
+
+from util.test_data_types import DeploymentResult
+
+
+class DeploymentRunner:
+    @abstractmethod
+    def run_deployment(self) -> DeploymentResult:
+        pass
```
Comment (filip-szweda) [inline roles/oneagent/tests/component/scripts/technology/deployment_runner.py:9]:
```suggestion
        raise NotImplementedError
```

File: roles/oneagent/tests/component/scripts/util/constant_key.py [hunk 33]
```diff
@@ -0,0 +1,5 @@
+from abc import abstractmethod
+
+
+def constant_key(func):
```
Comment (filip-szweda) [inline roles/oneagent/tests/component/scripts/util/constant_key.py:4]: Provide type hints.

File: roles/oneagent/tests/component/scripts/util/test_helpers.py [hunk 34]
```diff
@@ -0,0 +1,97 @@
...
+CallableOperation = Callable[[DeploymentPlatform, str, Any], None]
+
+
+def _get_param_by_name(name: str, **kwargs) -> Any:
+    assert name in kwargs, f"No '{name}' parameter in parameters list"
+    return kwargs[name]
+
+
+def disable_for_localhost():
+    def func_wrapper(func):
```
Comment (filip-szweda) [inline roles/oneagent/tests/component/scripts/util/test_helpers.py:23]: Provide type hints.

File: roles/oneagent/tests/component/scripts/util/test_helpers.py [hunk 35]
```diff
@@ -0,0 +1,97 @@
...
+                    return
+            func(*args, **kwargs)
+
+        return platforms_wrapper
+
+    return func_wrapper
+
+
+def enable_for_system_family(family: str):
+    def func_wrapper(func):
```
Comment (filip-szweda) [inline roles/oneagent/tests/component/scripts/util/test_helpers.py:38]: Provide type hints.

File: roles/oneagent/tests/component/scripts/util/test_helpers.py [hunk 36]
```diff
@@ -0,0 +1,97 @@
...
+                if any(host == "localhost" for host in hosts):
+                    return
+            func(*args, **kwargs)
+
+        return platforms_wrapper
+
+    return func_wrapper
+
+
+def enable_for_system_family(family: str):
```
Comment (filip-szweda) [inline roles/oneagent/tests/component/scripts/util/test_helpers.py:37]: Provide return type hint.

File: roles/oneagent/README.md [hunk 37]
```diff
@@ -53,6 +53,7 @@ The following variables are available in `defaults/main/` and can be overridden:
 | `oneagent_package_state` | `present` | OneAgent package state; use `present` or `latest` to make sure it's installed, or `absent` in order to uninstall.
 | `oneagent_reboot_host` | `false` | Reboot the secondary machine after OneAgent installation
 | `oneagent_validate_certs` | `true` | If set to `false`, allows to download OneAgent from a server with insecure SSL certificate (expired, self-signed, etc).
+| `oneagent_verify_signature` | `true` | Allows to skip verifying the installer's signature on UNIX platforms
```
Comment (jakubrak) [inline roles/oneagent/README.md:56]:
```suggestion
| `oneagent_verify_signature` | `true` | Verifies installer's signature (available only on AIX/Linux platforms)
```

File: roles/oneagent/defaults/main.yml [hunk 38]
```diff
@@ -31,5 +31,8 @@ oneagent_reboot_host: false
 # Set the need of validating certs during installer download
 oneagent_validate_certs: true
 
+# Set the need of verifying the installer signature
```
Comment (jakubrak) [inline roles/oneagent/defaults/main.yml:34]: comment is redundant (the same applies to other lines around)

File: roles/oneagent/tests/component/run.py [hunk 39]
```diff
@@ -0,0 +1,167 @@
...
+    for platform in [p.value for p in DeploymentPlatform]:
+        hosts = ','.join(args[platform])
+        hosts and test_args.append(f"--{platform}={hosts}")
+    return test_args
+
+
+def run_test(test: str, test_args: list[str]) -> bool:
+    test_name = Path(test).stem
+    logging.info(f"Test: {test_name}")
+    proc = subprocess.run(["pytest", test] + test_args, env=get_env_vars(), encoding="utf-8",
```
Comment (jakubrak) [inline roles/oneagent/tests/component/run.py:60]: please do not call pytest as subprocess. Pytest have all you need to configure and run the tests

File: roles/oneagent/tests/component/scripts/tests/test_installAndConfig.py [hunk 40]
```diff
@@ -0,0 +1,160 @@
...
+    perform_operation_on_platforms(platforms, _check_install_args, wrapper, constants.TECH_NAME)
+
+
+def test_oneagentctl_installation_config(_set_up, runner, configurator, platforms, wrapper):
+    logging.info("Running oneagentctl config test")
+
+    configurator.clear_parameters_section()
+    set_installer_download_params(configurator)
+    configurator.set_common_parameter(configurator.INSTALLER_ARGS_KEY, CONFIG_INSTALL_ARGS)
+    configurator.set_common_parameter(configurator.VERIFY_SIGNATURE_KEY, False)
```
Comment (jakubrak) [inline roles/oneagent/tests/component/scripts/tests/test_installAndConfig.py:126]: please cover signature verification. You can serve self-signed certificate in the same way as installer itself

File: .github/actions/run-tests/action.yaml [hunk 41]
```diff
@@ -20,7 +20,19 @@ runs:
...
+      shell: bash
+      run: python -m virtualenv venv && source venv/bin/activate && cd roles/oneagent/tests/component &&
+           pip install -r resources/requirements.txt && python run.py --linux_x86=localhost
+    - name: Upload logs
+      if: failure() && steps.component-tests.outcome == 'failure'
+      uses: actions/upload-artifact@v4
+      with:
+        name: component-tests-${{ github.sha }}
+        retention-days: 1
+        path: roles/oneagent/tests/component/test_dir/logs
```
[the remark this answers is not in the document]
Comment (WStechura) [reply .github/actions/run-tests/action.yaml:38]: Logs from deployment are pretty long. Combined with logs from tests are pretty unreadible

File: roles/oneagent/tests/component/scripts/command/windows/windows_command_executor.py [hunk 42]
```diff
@@ -0,0 +1,14 @@
...
+from util.test_data_types import CommandResult
+
+
+class WindowsCommandExecutor:
+    def __init__(self, user: str, password: str):
+        self.user = user
+        self.password = password
+
+    def execute(self, address: str, command: str, *args: str) -> CommandResult:
+        session = winrm.Session(address, auth=(self.user, self.password))
```
Comment (karczex) [inline roles/oneagent/tests/component/scripts/command/windows/windows_command_executor.py:12]: Here, the issue with session management is even better visible than in ssh case, as pywinrm sessions are designed to be maintained for some longer period of time.

File: roles/oneagent/tests/component/scripts/command/windows/windows_command_wrapper.py [hunk 43]
```diff
@@ -0,0 +1,37 @@
+from pathlib import Path
+
+from command.command_wrapper import CommandWrapper
+from command.windows.windows_command_executor import WindowsCommandExecutor
+from util.test_data_types import CommandResult
+
+
+class WindowsCommandWrapper(CommandWrapper):
```
Comment (karczex) [inline roles/oneagent/tests/component/scripts/command/windows/windows_command_wrapper.py:8]: Same as unix one

File: roles/oneagent/tests/component/scripts/server/server.key
Comment (karczex) [inline roles/oneagent/tests/component/scripts/server/server.key:1]: May it be generated in runtime?

File: roles/oneagent/tests/component/scripts/command/unix/unix_command_executor.py [hunk 44]
```diff
@@ -0,0 +1,36 @@
...
+        return ["sshpass", "-p", f"{password}", "ssh", "-o", "StrictHostKeyChecking=no", f"{user}@{address}"]
```
[the remark this answers is not in the document]
Comment (jakubrak) [reply roles/oneagent/tests/component/scripts/command/unix/unix_command_executor.py:10]: since for now those tests will run only on localhost, I think we can get rid of that code

File: roles/oneagent/tests/component/scripts/util/ssl_certificate_generator.py [hunk 45]
```diff
@@ -0,0 +1,78 @@
...
+        ]))
+        builder = builder.not_valid_before(datetime.utcnow())
+        builder = builder.not_valid_after(datetime.utcnow() + timedelta(days=self.validity_days))
+        builder = builder.serial_number(int(uuid.uuid4()))
+        builder = builder.public_key(self.public_key)
+        builder = builder.add_extension(
+            x509.SubjectAlternativeName([x509.DNSName(u"localhost")]),
+            critical=False,
+        )
+        self.certificate = builder.sign(
```
Comment (karczex) [inline roles/oneagent/tests/component/scripts/util/ssl_certificate_generator.py:54]: it seems `certificate` may be local variable of this method. You do not need to save its result as class state, just return cert. If called twice, cert will be calculated two times anyway. 

you may use it: 
```
self._save_certificate(self._generate_certificate())

File: roles/oneagent/tests/component/resources/installers/certificate.pem
[the remark this answers is not in the document]
Comment (jakubrak) [reply roles/oneagent/tests/component/resources/installers/certificate.pem:1]: removed

File: roles/oneagent/tests/component/scripts/command/unix/unix_command_executor.py [hunk 46]
```diff
@@ -0,0 +1,36 @@
...
+    prefix = _get_command_prefix(address, user, password)
+    # If ssh prefix is absent, the command is executed locally, so we need to use shell=True and make it string
+    # to avoid interpreting args as separate commands and avoid problems with non-existing commands
+    # If prefix exists, we do not need shell as we will be executing the command on the remote machine
+    if not prefix:
+        return " ".join([command, *args]), True
+    return prefix + [command, *args], False
+
+
+class UnixCommandExecutor:
```
[the remark this answers is not in the document]
Comment (jakubrak) [reply roles/oneagent/tests/component/scripts/command/unix/unix_command_executor.py]: I removed it completely since it's not used in tests

File: roles/oneagent/tests/conmponent/scripts/util/test_data_types.py (hunk 16 above)
Comment (jakubrak) [reply roles/oneagent/tests/conmponent/scripts/util/test_data_types.py]: Not sure what you mean here?

File: roles/oneagent/tests/component/scripts/command/unix/unix_command_wrapper.py [hunk 47]
```diff
@@ -0,0 +1,26 @@
...
+
+class UnixCommandWrapper(CommandWrapper):
+    def __init__(self, user: str, password: str):
+        self.password = password
+        self.executor = UnixCommandExecutor(user, password)
+
+    def get_file_content(self, address: str, file: Path) -> CommandResult:
+        return self.executor.execute(address, "cat", str(file))
+
+    def file_exists(self, address: str, file: Path) -> CommandResult:
```
[the remark this answers is not in the document]
Comment (jakubrak) [reply roles/oneagent/tests/component/scripts/command/unix/unix_command_wrapper.py]: Agree, added to backlog

File: roles/oneagent/tests/conmponent/scripts/tests/test_resilience.py [hunk 48]
```diff
@@ -0,0 +1,166 @@
...
+
+    set_installer_download_params(configurator)
+    configurator.set_common_parameter(configurator.INSTALLER_ARCH_KEY, "unknown_arch")
+
+    _check_deployment_failure(
+        run_deployment(runner, True), _error_messages[UNKNOWN_ARCHITECTURE_KEY], constants.FAILED_DEPLOYMENT_EXIT_CODE
+    )
+
+
+# noinspection PyUnusedLocal
```
[the remark this answers is not in the document]
Comment (jakubrak) [reply roles/oneagent/tests/conmponent/scripts/tests/test_resilience.py]: pylint fixes added to the backlog

File: roles/oneagent/tests/conmponent/scripts/util/test_helpers.py (hunk 12 above)
Comment (jakubrak) [reply roles/oneagent/tests/conmponent/scripts/util/test_helpers.py]: let's move it to the backlog

**Review by jakubrak — APPROVED** [review]

File: README.md [hunk 49]
```diff
@@ -5,7 +5,7 @@ In its current state, the collection consists of a single role that deploys Dyna
 ### General
 * Ansible >= 2.15.0
 ### Windows
-* pywinrm >= 0.4.1
+* pywinrm >= 0.4.3
```
[the remark this answers is not in the document]
Comment (jakubrak) [reply README.md]: Windows testing is currently not supported
